### ISSUE TITLE
Recursive project delete

### ISSUE THREAD

**Issue by david-crespo** [oxidecomputer/omicron#4674]
By design, delete operations on any resource fail when the resource has children. In order to delete a whole project containing instances, disks, VPCs, etc., one currently has to build the tree of resource and delete everything from the leaves up. The problem is: that is often something we want to do! Currently we have some scripts that do this client-side. That may be where we land (maybe, for example, incorporating this logic into the CLI) but I wanted to track the possibility of doing this within the API. It would be a pretty complex and slow operation involved stopping potentially many instances and deleting many resources. It would probably make sense to wait for #392.

**Comment by davepacheco** [comment]
Yeah, I can definitely see the appeal.  One thing I worry about is that it breaks the usual rule of doing O(1) work per operation, which helps protect against DoS, intentional or otherwise.  Someone could construct a gigantic project, then with very little extra work, kick off this potentially very expensive operation for us without them having to do any work to keep it going.  We could potentially mitigate this by throttling the delete itself so that at least one giant delete can't swamp the system.

**Comment by benjaminleonard** [comment]
A while back I was working on designs for a delete modal that would show within it the tree of child resources and whether they would be deleted or not — e.g. when an instance deleted, showing whether the associated disks are deleted too or just detached.

Perhaps a reasonable middle-ground would be an endpoint that returned that tree instead. A list of resources and whether they would also be deleted. We could then do that action async on the console. Or at least batch the deletes a few at a time, with a realtime look at the progress of that action, no unwinding necessary.

This implementation would also mean that we can let a user know when they go to delete a project (or other resource) why they are unable to client-side. Without just knocking them back with an error, which is a less than ideal user experience.

**Comment by davepacheco** [comment]
> Perhaps a reasonable middle-ground would be an endpoint that returned that tree instead. A list of resources and whether they would also be deleted. We could then do that action async on the console. Or at least batch the deletes a few at a time, with a realtime look at the progress of that action, no unwinding necessary.

How would this differ from the paginated interfaces that we have today for listing everything in the project?  Now that you mention it: if we do build "recursive delete" entirely on the client, then the client would wind up assembling a list of everything that has to be deleted (using our paginated (i.e., safe) interfaces) and could present that to the user for confirmation before deleting.

I'm not opposed to a server-side implementation either but we'd want to be very careful about implementing it in such a way that it's async, heavily throttled, and continues running even if Nexus instances go away or even get destroyed, so it's a bit of work.

**Comment by benjaminleonard** [comment]
> How would this differ from the paginated interfaces that we have today for listing everything in the project? 

I suppose presently you have to hit each endpoint of any child resource type, the idea would be you'd get in a single response a list of everything that has to be deleted. 

It might not make the implementation of delete any better – since you have to call the respective endpoint for each to delete anyway. But it would make it easy to present the list of items in a project to an end-user trying to delete the project.
